flow/src/main/java/flow/History.java

```diff
@@ -173,32 +180,49 @@ public int hashCode() {
 
   public static final class Builder {
     private final Deque<Entry> history;
+    private final Map<Object, Entry> entryMemory = new LinkedHashMap<>();
 
     private Builder(Collection<Entry> history) {
       this.history = new ArrayDeque<>(history);
     }
 
+    public Builder clear() {
+      while (!history.isEmpty()) {
+        pop();
+      }
+      return this;
+    }
+
     public Builder push(Object object) {
-      Entry entry = new Entry(object);
+      Entry entry = entryMemory.get(object);
+      if (entry == null) {
+        entry = new Entry(object);
+      }
       history.push(entry);
-
+      entryMemory.remove(object);
       return this;
     }
 
     public Builder addAll(Collection<Object> c) {
       for (Object object : c) {
-        history.push(new Entry(object));
+        push(object);
       }
 
       return this;
     }
 
+    Entry peekEntry() {
```

Comment:
Weird that you extract this method but used it in only one place. I'd inline, or if you really want it should move private methods below public ones.